app/Platform/Services/AchievementSetClaimListService.php

```diff
@@ -0,0 +1,449 @@
+                break;
+
+            case 'enddate':
+                $claims->orderBy('Finished');
+                break;
+
+            case '-expiring':
+                $claims->orderByRaw(ifStatement("SetClaim.Status IN(" . ClaimStatus::Active . ',' . ClaimStatus::InReview . ")", 0, 1))
+                       ->orderBy('Finished');
+                break;
+
+            case 'expiring':
+                $claims->orderByRaw(ifStatement("SetClaim.Status IN(" . ClaimStatus::Active . ',' . ClaimStatus::InReview . ")", 0, 1))
+                       ->orderByDesc('Finished');
+                break;
+        }
+
+        return $claims->with(['game.system', 'user']);
+    }
+
+    public function getGameColumn(): array
+    {
+        return [
+            'header' => 'Game',
+            'render' => function ($claim) {
+                echo Blade::render('
+                    <x-game.multiline-avatar
+                        :gameId="$ID"
+                        :gameTitle="$Title"
+                        :gameImageIcon="$ImageIcon"
+                        :consoleName="$consoleName"
+                    />', [
+                        'ID' => $claim->game->id,
+                        'Title' => $claim->game->title,
+                        'ImageIcon' => $claim->game->ImageIcon,
+                        'consoleName' => $claim->game->system->name,
+                    ]);
+            },
+        ];
+    }
+
+    public function getDeveloperColumn(): array
+    {
+        return [
+            'header' => 'Developer',
+            'render' => function ($claim) {
+                echo userAvatar($claim->user);
+            },
+        ];
+    }
+
+    public function getClaimTypeColumn(): array
+    {
+        return [
+            'header' => 'Claim Type',
+            'render' => function ($claim) {
+                echo ClaimType::toString($claim->ClaimType);
+            },
+        ];
+    }
+
+    public function getSetTypeColumn(): array
+    {
+        return [
+            'header' => 'Set Type',
+            'render' => function ($claim) {
+                echo ClaimSetType::toString($claim->SetType);
+            },
+        ];
+    }
+
+    public function getStatusColumn(): array
+    {
+        return [
+            'header' => 'Status',
+            'render' => function ($claim) {
+                echo ClaimStatus::toString($claim->Status);
+            },
+        ];
+    }
+
+    public function getSpecialColumn(): array
+    {
+        return [
+            'header' => 'Special',
+            'render' => function ($claim) {
+                echo ClaimSpecial::toString($claim->Special);
+            },
+        ];
+    }
+
+    private function renderDate(?Carbon $date): void
+    {
+        echo '<span class="smalldate whitespace-nowrap">';
+        if ($date) {
+            echo getNiceDate($date->unix());
+        } else {
+            echo 'Unknown';
+        }
+        echo '</span>';
+    }
```

Comment:
If we can get away with it, I can't help but feel it's strongly preferably for our services to not `echo` HTML. This blurs the line between our application logic and our view logic.

What might be more fitting is for the service to communicate that we want to render a date (something like "this cell is a date"), and the template having a conditional where if given that message of "this cell is a date", renders the HTML.

Same feedback applies for expiration status and game.

Reply:
I agree that defining a unique view for each combination of columns that gets displayed is also undesirable.

My feedback for those other two examples is the same. `renderDate()` itself is a component masquerading as a function.

The best way I can think of in this case to separate the presentation and business logic is to return data arrays in the `render` field, like:

```php
public function getExpirationDateColumn(): array
{
    return [
        'header' => 'Expires At',
        'render' => function ($claim) {
            return [
                'type' => 'date',
                'value' => $claim->Finished,
            ];
        },
    ];
}

public function getExpirationStatusColumn(): array
{
    return [
        'header' => 'Expiration Status',
        'render' => function ($claim) {
            return [
                'type' => 'status',
                'status' => $claim->Status,
                'finished' => $claim->Finished,
            ];
        },
    ];
}
```

And then fully delegate the presentation details to the component with conditions against the `type` value:

```php
@if ($column['render]['type'] === 'date')
  <!-- some html -->
@elseif ($column['render']['type'] === 'status')
  <!-- some html -->
@endif
```

This provides several tangible and concrete benefits:
* Business logic is now isolated to the service layer, and presentation logic is isolated in the view layer. Each layer now has a single responsibility and well-defined role.
* Changes to the presentation can now be made without needing to dive into or alter the business logic, and vice versa.
* It's easier to write tests for business logic that doesn't depend on presentation details. Similarly, if we wanted to do UI tests in the future, the view can be tested independently of its logic.
* The data returned by the service can actually be reused in different contexts or views without being tied to a specific presentation format.

Reply:
done